File: MazeMaker.java
import java.util.*;

public class MazeMaker {
  public int longestPath(String[] maze, int startRow, int startCol, int[] moveRow, int[] moveCol) {

    Queue<int[]> queue = new LinkedList<>();
    int h = maze.length;
    int w = maze[0].length();
    int[][] mins = new int[h][w];
    for (int i = 0; i < h; i++) {
      for (int j = 0; j < w; j++) {
        mins[i][j] = Integer.MAX_VALUE;
      }
    }
    mins[startRow][startCol] = 0;

    queue.add(new int[] { startRow, startCol, 0 });

    while (!queue.isEmpty()) {
      int[] values = queue.poll();
      int y = values[0];
      int x = values[1];
      int c = values[2];

      for (int mi = 0; mi < moveRow.length; mi++) {
        int ny = y + moveRow[mi];
        int nx = x + moveCol[mi];

        if (ny < 0 || ny >= h || nx < 0 || nx >= w)
          continue;

        if (maze[ny].charAt(nx) != '.')
          continue;

        if (c + 1 >= mins[ny][nx])
          continue;

        mins[ny][nx] = c + 1;

        queue.add(new int[] { ny, nx, c + 1 });
      }
    }

    int ans = 0;
    for (int i = 0; i < h; i++) {
      for (int j = 0; j < w; j++) {
        if (maze[i].charAt(j) != '.')
          continue;

        if (mins[i][j] == Integer.MAX_VALUE) {
          return -1;
        } else {
          ans = Math.max(ans, mins[i][j]);
        }
      }
    }

    return ans;
  }

  public static void main(String[] args) {
    MazeMaker mm = new MazeMaker();
    System.out.println(
        mm.longestPath(new String[] { "..X.X.X.X.X.X." }, 0, 0, new int[] { 2, 0, -2, 0 }, new int[] { 0, 2, 0, -2 }));
  }
}

File: MazeMaker.cpp
#include <vector>
#include <algorithm>
#include <iostream>
#include <queue>

using namespace std;

class MazeMaker
{
public:
  int longestPath(vector<string> maze, int startRow, int startCol, vector<int> moveRow, vector<int> moveCol)
  {
    const int INF = 100000000;
    int h = maze.size();
    int w = maze[0].size();
    int mins[h][w];
    for (int i = 0; i < h; i++)
    {
      for (int j = 0; j < w; j++)
      {
        mins[i][j] = INF;
      }
    }

    queue<int> qRow;
    queue<int> qCol;
    queue<int> qCost;

    qRow.push(startRow);
    qCol.push(startCol);
    qCost.push(0);
    mins[startRow][startCol] = 0;

    while (!qRow.empty())
    {
      int row = qRow.front();
      int col = qCol.front();
      int cost = qCost.front();
      qRow.pop();
      qCol.pop();
      qCost.pop();

      for (int mi = 0; mi < moveRow.size(); mi++)
      {
        int nRow = row + moveRow[mi];
        int nCol = col + moveCol[mi];

        if (nRow < 0 || nRow >= h || nCol < 0 || nCol >= w)
        {
          continue;
        }
        if (maze[nRow][nCol] == 'X')
        {
          continue;
        }
        if (mins[nRow][nCol] <= cost + 1)
        {
          continue;
        }

        mins[nRow][nCol] = cost + 1;
        qRow.push(nRow);
        qCol.push(nCol);
        qCost.push(cost + 1);
      }
    }

    int ans = 0;
    for (int i = 0; i < h; i++)
    {
      for (int j = 0; j < w; j++)
      {
        // cout << mins[i][j] << " ";
        if (maze[i][j] == 'X')
        {
          continue;
        }
        if (mins[i][j] == INF)
        {
          return -1;
        }
        else
        {
          ans = max(ans, mins[i][j]);
        }
      }
      // cout << endl;
    }
    return ans;
  }
};

int main()
{
  MazeMaker *c = new MazeMaker;
  cout << c->longestPath({"......."}, 0, 0, {1, 0, 1, 0, 1, 0}, {0, 1, 0, 1, 0, 1}) << endl;
}